SEGV when creating 14GB zar archive

I hit another bug when attempting to create an archive similar to the 14GB conn.log I used for issue #1758 .
The archive for 1758 was created with -s 100GB, but for this time, I left off the -s parameter so it would default to 500GB chunks.

On Matt's advice, I set the -bufsize large so it would consistently create large chunks. Instead of finishing after 94 minutes, like 1758, it SEGV's after 84 minutes. I will try again with -bufsize 2GB.

Here is the stack trace:

[ec2-user@ip-172-31-19-39 ~]$ time zar import -bufsize 1000MB -R s3://zqd-demo-1/mark/zeek-logs/conn s3://brim-sampledata/wrccdc/zeek-logs/conn.log.gz
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xa44e22]

goroutine 42 [running]:
github.com/brimsec/zq/proc/spill.(*MergeSort).Cleanup(0x0)
	/home/ec2-user/zq/proc/spill/merge.go:53 +0x22
panic(0xd2c400, 0x16e2010)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/brimsec/zq/proc/spill.(*MergeSort).Len(...)
	/home/ec2-user/zq/proc/spill/merge.go:119
github.com/brimsec/zq/proc/spill.(*MergeSort).Read(0x0, 0x0, 0x0, 0x0)
	/home/ec2-user/zq/proc/spill/merge.go:94 +0x26
github.com/brimsec/zq/zbuf.CopyWithContext(0x10b1820, 0xc0000c8b40, 0x109ade0, 0xc0009ce000, 0x109afa0, 0x0, 0x0, 0xc0006e2065)
	/home/ec2-user/zq/zbuf/zbuf.go:59 +0x58
github.com/brimsec/zq/ppl/archive.(*tsDirWriter).flush(0xc0fe4b21e0, 0x0, 0x0)
	/home/ec2-user/zq/ppl/archive/import.go:335 +0x341
github.com/brimsec/zq/ppl/archive.(*Writer).flushStaleWriters(0xc000594680, 0xc000063ec8, 0x2)
	/home/ec2-user/zq/ppl/archive/import.go:159 +0x30b
github.com/brimsec/zq/ppl/archive.(*Writer).flusher(0xc000594680, 0x0, 0x0)
	/home/ec2-user/zq/ppl/archive/import.go:135 +0x12f
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc0002b6ea0, 0xc000658620)
	/home/ec2-user/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:57 +0x59
created by golang.org/x/sync/errgroup.(*Group).Go
	/home/ec2-user/go/pkg/mod/golang.org/x/sync@v0.0.0-20190911185100-cd5d95a43a6e/errgroup/errgroup.go:54 +0x66

real	84m5.390s
user	108m56.378s
sys	2m18.821s

--- a/ppl/archive/import.go
+++ b/ppl/archive/import.go
@@ -55,10 +55,10 @@ type Writer struct {
 	cancel        context.CancelFunc
 	ctx           context.Context
 	errgroup      *errgroup.Group
+	mu            sync.Mutex
 	once          sync.Once
 	staleDuration time.Duration
 	writers       map[tsDir]*tsDirWriter
-	writersMu     sync.RWMutex
 
 	memBuffered int64
 	stats       ImportStats
@@ -96,27 +96,26 @@ func (w *Writer) Write(rec *zng.Record) error {
 		return w.ctx.Err()
 	}
 
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
 	tsDir := newTsDir(rec.Ts())
-	w.writersMu.RLock()
 	dw, ok := w.writers[tsDir]
-	w.writersMu.RUnlock()
 	if !ok {
 		var err error
 		dw, err = newTsDirWriter(w, tsDir)
 		if err != nil {
 			return err
 		}
-		w.writersMu.Lock()
 		if _, ok := w.writers[tsDir]; !ok {
 			w.writers[tsDir] = dw
 		}
 		dw = w.writers[tsDir]
-		w.writersMu.Unlock()
 	}
 	if err := dw.Write(rec); err != nil {
 		return err
 	}
-	for atomic.LoadInt64(&w.memBuffered) > ImportBufSize {
+	for w.memBuffered > ImportBufSize {
 		if err := w.spillLargestBuffer(); err != nil {
 			return err
 		}
@@ -145,16 +144,18 @@ func (w *Writer) flusher() error {
 // removes any writers that haven't recieved any updates since
 // Writer.staleDuration.
 func (w *Writer) flushStaleWriters() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
 	now := time.Now()
-	w.writersMu.Lock()
 	var stale []*tsDirWriter
 	for tsd, writer := range w.writers {
-		if now.Sub(writer.modified()) >= w.staleDuration {
+		if now.Sub(writer.modts.Time()) >= w.staleDuration {
 			stale = append(stale, writer)
 			delete(w.writers, tsd)
 		}
 	}
-	w.writersMu.Unlock()
+
 	for _, writer := range stale {
 		if err := writer.flush(); err != nil {
 			return err
@@ -168,8 +169,6 @@ func (w *Writer) flushStaleWriters() error {
 // spilling to disk the records of the tsDirWriter with the largest memory
 // footprint.
 func (w *Writer) spillLargestBuffer() error {
-	w.writersMu.RLock()
-	defer w.writersMu.RUnlock()
 	var largest *tsDirWriter
 	for _, dw := range w.writers {
 		if largest == nil || dw.bufSize > largest.bufSize {
@@ -180,7 +179,7 @@ func (w *Writer) spillLargestBuffer() error {
 }
 
 func (w *Writer) Close() error {
-	w.writersMu.Lock()
+	w.mu.Lock()
 	var merr error
 	for ts, dw := range w.writers {
 		if err := dw.flush(); err != nil {
@@ -188,7 +187,7 @@ func (w *Writer) Close() error {
 		}
 		delete(w.writers, ts)
 	}
-	w.writersMu.Unlock()
+	w.mu.Unlock()
 	w.cancel()
 	if err := w.errgroup.Wait(); merr == nil {
 		merr = err
@@ -228,7 +227,7 @@ type tsDirWriter struct {
 	ark     *Archive
 	bufSize int64
 	ctx     context.Context
-	modts   int64
+	modts   nano.Ts
 	records []*zng.Record
 	spiller *spill.MergeSort
 	tsDir   tsDir
@@ -250,7 +249,7 @@ func newTsDirWriter(w *Writer, tsDir tsDir) (*tsDirWriter, error) {
 
 func (dw *tsDirWriter) addBufSize(delta int64) {
 	dw.bufSize += delta
-	atomic.AddInt64(&dw.writer.memBuffered, delta)
+	dw.writer.memBuffered += delta
 }
 
 // chunkSizeEstimate returns a crude approximation of all records when written
@@ -281,11 +280,7 @@ func (dw *tsDirWriter) Write(rec *zng.Record) error {
 }
 
 func (dw *tsDirWriter) touch() {
-	atomic.StoreInt64(&dw.modts, int64(nano.Now()))
-}
-
-func (dw *tsDirWriter) modified() time.Time {
-	return nano.Ts(atomic.LoadInt64(&dw.modts)).Time()
+	dw.modts = nano.Now()
 }
 
 func (dw *tsDirWriter) spill() error {

